### ISSUE TITLE
Bump patch version when a new checklist is created, not when it's locked

### ISSUE THREAD

**Issue by roryabraham** [Expensify/App#7166]
cc @Expensify/mobile-deployers

### Problem 1
It is possible for NewDot CI/CD to get into a state where all mobile staging deploys are temporarily failing. This can happen when:

1. A checklist is closed, and a version (such as `1.1.27-1`) is deployed to production.
1. Before the new checklist is locked, the production deploy passes review and ships for sale. Then the `1.1.27` track is closed.
1. Staging deploys that happen after this fail, until the new checklist is locked and `1.1.28-0` is created.

### Why it's important
There's a couple reasons one could argue this isn't _really_ a problem:

1. All the code on main will be deployed to staging when the checklist is locked, so what does it matter if deploys fail in the meantime?
2. It's sort of rare because, generally speaking, the new checklist is locked in less than 24 hours (the average turnaround time for a production deploy).

However, I think there a handful of pretty compelling reasons to solve this:

1. The above-listed reasons for why it's _not_ a problem aren't very obvious and require some "tribal knowledge".
1. This causes red-herring deploy failures and slack noise that waste deployer time.
1. This makes it so that newly-merged PRs won't deploy to staging until the checklist is locked, so anyone trying to do QA before Applause locks the checklist will have to wait.
1. I think the solution laid out below is much more intuitive and will make the deploy process easier to understand and explain. Each deploy checklist gets it's own `PATCH` version from the get-go.
1. When non-deployers see some platform failures on their deploy comments, they may get confused or waste time investigating if their PR broke something.

---

### Problem 2
There exists the following race condition:

1. Lock the deploy checklist.
1. The `lockDeploys.yml` workflow starts creating a new `PATCH` version.
1. Then immediately merge another PR to main.
1. Then once the new `PATCH` version is created, `main` is merged into `staging`, and the new PR is included in the staging deploy.

### Why it's important
While it's not a hugely important problem (and happens only rarely), it creates some unexpected behavior:

1. The merged PR will get a "This PR is not deployed to staging because the checklist is locked" comment, but that's wrong. 
1. When it is deployed to staging, it does not get a "deployed to staging" comment, and it will appear as if it was somehow deployed directly to production.
1. Also, it's just not intuitive that a PR merged _after_ the deploy checklist is locked would be deployed to staging.

---

### Solution (for both)
Change the deploy/semver paradigm a little bit, such that we bump the `PATCH` version on `main`/`staging` when a new checklist is created, _not_ when the checklist is locked.

Some caveats of this solution:

- We'll no longer need to run a staging deploy at the time the checklist locks. Hooray for less noise and less waiting for deploys to finish! However, we'll need to train Applause to:
    1. Lock the checklist when they're ready to start QA
    1. Make sure that they check for any ongoing staging deploys, and if there are any running, wait for those to finish before they actually start QA.
        - **Note:** This is technically something they need to do in our existing paradigm, so this isn't a "breaking change". But we'll want to reiterate this to them because the deploy that currently happens when a checklist is locked is redundant (only code change included is the `PATCH` version bump), so it's less likely that code will be missing when they start QA.
- We'll need to make a ton of documentation changes:
    - [ ] StackOverflow posts explaining the deploy processes
    - [x] README.md's covering deploy processes
    - [ ] In-code documentation
    - [x] Unit test changes in `tests/unit/getPullRequestsMergedBetween.sh`
    - [ ] Announce these changes to applause and mobile deployers
    - [ ] Probably others that I'm forgetting about

**Comment by roryabraham** [comment]
Got a draft PR up for this 👍
